confidential-data-hub/hub/src/bin/grpc_server/mod.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::*;
+
+use confidential_data_hub::{hub::Hub, DataHub};
+use log::{debug, error};
+use std::{error::Error as _, net::SocketAddr, sync::Arc};
+use storage::volume_type::Storage;
+use tokio::sync::RwLock;
+use tonic::{transport::Server, Request, Response, Status};
+
+use crate::message::{KeyProviderInput, KeyUnwrapOutput, KeyUnwrapResults};
+use api::{
+    get_resource_service_server::{GetResourceService, GetResourceServiceServer},
+    key_provider_service_server::{KeyProviderService, KeyProviderServiceServer},
+    sealed_secret_service_server::{SealedSecretService, SealedSecretServiceServer},
+    secure_mount_service_server::{SecureMountService, SecureMountServiceServer},
+    GetResourceRequest, GetResourceResponse, KeyProviderKeyWrapProtocolInput,
+    KeyProviderKeyWrapProtocolOutput, SecureMountRequest, SecureMountResponse, UnsealSecretInput,
+    UnsealSecretOutput,
+};
+
+mod api {
+    tonic::include_proto!("api");
+    tonic::include_proto!("keyprovider");
+}
+
+macro_rules! format_error {
+    ($err:expr) => {{
+        let mut error_string = format!("{}", $err);
+        let mut current_error = $err.source();
+        while let Some(source) = current_error {
+            error_string.push_str(&format!("\nCaused by: {}", source));
+            current_error = source.source();
+        }
+        error_string
+    }};
+}
+
+pub struct Cdh {
+    inner: RwLock<Hub>,
+}
+
+#[tonic::async_trait]
+impl SealedSecretService for Arc<Cdh> {
+    async fn unseal_secret(
+        &self,
+        request: Request<UnsealSecretInput>,
+    ) -> Result<Response<UnsealSecretOutput>, Status> {
+        debug!("get new UnsealSecret request");
+        let request = request.into_inner();
+
+        let cdh = self.inner.read().await;
+
+        let plaintext = cdh.unseal_secret(request.secret).await.map_err(|e| {
+            let detailed_error = format_error!(e);
+            error!("Call CDH to unseal secret failed:\n{detailed_error}");
+            Status::internal(format!("[ERROR] CDH unseal secret failed: {}", e))
```

Comment:
Maybe these errors should also mention that it is gRPC CDH like we do for the AA?